api/secret.go

```diff
@@ -35,6 +37,268 @@ type Secret struct {
+		return nil
+	}
+
+	metaRaw, ok := s.Data["meta"].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+
+	meta := make(map[string]string, len(metaRaw))
+	for k, v := range metaRaw {
+		m, ok := v.(string)
+		if !ok {
+			return nil
+		}
+		meta[k] = m
+	}
+
+	return meta
+}
+
+// TokenRemainingUses returns the standardized remaining uses for the given
+// secret. If the secret is nil or does not contain the "num_uses", this returns
+// 0..
+func (s *Secret) TokenRemainingUses() int {
+	if s == nil || s.Data == nil || s.Data["num_uses"] == nil {
+		return 0
+	}
+
+	usesStr, ok := s.Data["num_uses"].(json.Number)
+	if !ok {
+		return 0
+	}
+
+	if string(usesStr) == "" {
+		return 0
+	}
+
+	uses, err := strconv.ParseInt(string(usesStr), 10, 64)
+	if err != nil {
+		return 0
+	}
+
+	return int(uses)
+}
+
+// TokenPolicies returns the standardized list of policies for the given secret.
+// If the secret is nil or does not contain any policies, this returns nil.
+// Policies are usually returned as []interface{}, but this function ensures
+// they are []string.
+func (s *Secret) TokenPolicies() []string {
+	if s == nil {
+		return nil
+	}
+
+	if s.Auth != nil && len(s.Auth.Policies) > 0 {
+		return s.Auth.Policies
+	}
+
+	if s.Data == nil || s.Data["policies"] == nil {
+		return nil
+	}
+
+	list, ok := s.Data["policies"].([]interface{})
+	if !ok {
+		return nil
+	}
+
+	policies := make([]string, len(list))
+	for i := range list {
+		p, ok := list[i].(string)
+		if !ok {
+			return nil
+		}
+		policies[i] = p
+	}
+
+	return policies
+}
+
+// TokenMetadata returns the map of metadata associated with this token, if any
+// exists. If the secret is nil or does not contain the "metadata" key, this
+// returns nil.
+func (s *Secret) TokenMetadata() map[string]string {
+	if s == nil {
+		return nil
+	}
+
+	if s.Auth != nil && len(s.Auth.Metadata) > 0 {
+		return s.Auth.Metadata
+	}
+
+	if s.Data == nil || (s.Data["metadata"] == nil && s.Data["meta"] == nil) {
+		return nil
+	}
+
+	data, ok := s.Data["metadata"].(map[string]interface{})
+	if !ok {
+		data, ok = s.Data["meta"].(map[string]interface{})
+		if !ok {
+			return nil
+		}
+	}
+
+	metadata := make(map[string]string, len(data))
+	for k, v := range data {
+		typed, ok := v.(string)
+		if !ok {
+			return nil
+		}
+		metadata[k] = typed
+	}
+
+	return metadata
+}
+
+// TokenIsRenewable returns the standardized token renewability for the given
+// secret. If the secret is nil or does not contain the "renewable" key, this
+// returns false.
+func (s *Secret) TokenIsRenewable() bool {
+	if s == nil {
+		return false
+	}
+
+	if s.Auth != nil && s.Auth.Renewable {
+		return s.Auth.Renewable
+	}
+
+	if s.Data == nil || s.Data["renewable"] == nil {
+		return false
+	}
+
+	renewable, ok := s.Data["renewable"].(bool)
+	if !ok {
+		return false
+	}
+
+	return renewable
+}
+
+// TokenTTLInt returns the token's TTL as an integer number of seconds.
+func (s *Secret) TokenTTLInt() int {
```

Comment:
I would get rid of this in favor of the call below. time.Duration is a standard way to do durations in Go code and anyone that needs the number of seconds can get it directly from that.

Reply:
This is here because LeaseDuration is always specified as seconds and often stored as an `int` in some places. time.Duration.Seconds() returns a float, which then has to be rounded and converted to seconds on each call.